python/ql/lib/semmle/python/security/dataflow/ServerSideRequestForgeryCustomizations.qll

```diff
@@ -137,4 +138,25 @@ module ServerSideRequestForgery {
       )
     }
   }
+
+  /** A validation that a string does not contain certain characters, considered as a sanitizer. */
+  private class StringRestrictionSanitizerGuard extends Sanitizer {
+    StringRestrictionSanitizerGuard() {
+      this = DataFlow::BarrierGuard<stringRestriction/3>::getABarrierNode()
+    }
+  }
+
+  private predicate stringRestriction(DataFlow::GuardNode g, ControlFlowNode node, boolean branch) {
+    exists(DataFlow::MethodCallNode call, DataFlow::Node strNode |
+      call.asCfgNode() = g and strNode.asCfgNode() = node
+    |
+      branch = true and
+      call.calls(strNode,
+        ["isalnum", "isalpha", "isdecimal", "isdigit", "isidentifier", "isnumeric", "isspace"])
+      or
+      branch = true and
+      call = API::moduleImport("re").getMember(["match", "fullmatch"]).getACall() and
+      strNode = [call.getArg(1), call.getArgByName("string")]
```

Comment:
This is great, but I think we should also support the case where the regular expression is compiled in advance. That is

```python
import re
sanitizer_re = re.compile("...")

if sanitizer_re.match(input):
   # do stuff
```

Unfortunately, I don't think we expose such a notion in the libraries already (there's `compiledRegex` in `Stdlib.qll`, but it's in the internal `StdlibPrivate` module). However, I think in this case adding a separate case for `API::moduleImport("re").getMember("compile").getReturn()` (and then the methods you include above) should suffice. (Note that the argument position changes slightly for compiled regexes, so it's not simply a case of augmenting the `API::moduleImport("re")` bit above.)